File: src/com/backyardbrains/usb/SampleStreamProcessor.java
package com.backyardbrains.usb;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.SparseArray;
import com.backyardbrains.audio.Filters;
import com.backyardbrains.data.processing.DataProcessor;
import com.backyardbrains.utils.SampleStreamUtils;
import com.backyardbrains.utils.SpikerBoxHardwareType;
import java.util.Arrays;

import static com.backyardbrains.utils.LogUtils.LOGD;
import static com.backyardbrains.utils.LogUtils.LOGW;
import static com.backyardbrains.utils.LogUtils.makeLogTag;

/**
 * @author Tihomir Leka <ticapeca at gmail.com>
 */
class SampleStreamProcessor implements DataProcessor {

    private static final String TAG = makeLogTag(SampleStreamProcessor.class);

    private static final int CLEANER = 0xFF;
    private static final int REMOVER = 0x7F;

    private static final int DEFAULT_CHANNEL_COUNT = 1;
    private static final int CHANNEL_INDEX = 0;

    // Responsible for detecting and processing escape sequences within incoming data
    private final EscapeSequence escapeSequence = new EscapeSequence();

    // Most recent unfinished frame
    private Frame unfinishedFrame;
    // Most recent unfinished sample
    private Sample unfinishedSample;
    // Holds currently processed channel
    private int currentChannel = 0;
    // Additional filtering that should be applied
    private Filters filters;
    // Number of channels
    private int channelCount = DEFAULT_CHANNEL_COUNT;
    // Whether channel count has changed during processing of the latest chunk of incoming data
    private boolean channelCountChanged;
    // Average signal which we use to avoid signal offset
    private double average;
    // Collection of events found within one sample batch
    private SparseArray<String> events;

    /**
     * Listens for responses sent by connected device as a response to custom messages sent by the application.
     */
    interface SampleStreamListener {
        /**
         * Triggered when SpikerBox sends hardware type message as a result of inquiry.
         */
        void onSpikerBoxHardwareTypeDetected(@SpikerBoxHardwareType int hardwareType);

        /**
         * Triggered when SpikerBox sends max sample rate and number of channels message as a result of inquiry.
         */
        void onMaxSampleRateAndNumOfChannelsReply(int maxSampleRate, int channelCount);
    }

    private SampleStreamListener listener;

    SampleStreamProcessor(@Nullable SampleStreamListener listener, @Nullable Filters filters) {
        this.listener = listener;
        this.filters = filters;
    }

    @NonNull @Override public short[] process(@NonNull byte[] data, @NonNull SparseArray<String> events) {
        if (data.length > 0) return processIncomingData(data, events);

        return new short[0];
    }

    /**
     * Set number of channels in the sample stream.
     */
    void setChannelCount(int channelCount) {
        this.channelCount = channelCount;

        channelCountChanged = true;
    }

    @NonNull private short[] processIncomingData(@NonNull byte[] data, @NonNull SparseArray<String> events) {
        this.events = events;

        // if channel count has changed during processing  previous data chunk we should disregard
        if (channelCountChanged) {
            unfinishedFrame = null;
            unfinishedSample = null;
            currentChannel = 0;

            channelCountChanged = false;
        }

        final int tmpChannelCount = channelCount;
        short[][] channels = new short[tmpChannelCount][];
        for (int i = 0; i < channels.length; i++) {
            // max number of samples can be number of incoming bytes divided by 2
            channels[i] = new short[(int) (data.length * .5)];
        }
        // array of sample counters, one for every channel
        int[] sampleCounters = new int[tmpChannelCount];
        int channelCounter = currentChannel;
        double avg = average;
        int lsb, msb; // less significant and most significant bytes
        short sample;

        for (byte b : data) {
            // 1. check if we are inside escape sequence or not
            // test the next byte to see if the sequence is valid
            if (!escapeSequence.test(b)) {
                //LOGD(TAG, "Escape sequence test failed!!");
                byte[] sequence = escapeSequence.getSequence();
                for (byte b1 : sequence) {
                    // check if we have unfinished frame
                    if (unfinishedFrame != null) {
                        // check if we have unfinished sample
                        if (unfinishedSample != null) {
                            lsb = b1 & CLEANER;

                            // if less significant byte is also grater then 127 drop whole frame
                            if (lsb > 127) {
                                LOGW(TAG, "LSB > 127! DROP WHOLE FRAME!");
                                unfinishedFrame = null;
                                unfinishedSample = null;
                                channelCounter = 0;
                                continue;
                            }

                            //LOGD(TAG, " --> LSB " + (channelCounter + 1) + ". CHANNEL");
                            unfinishedSample.setLsb(lsb);
                            if (!unfinishedFrame.isFull()) unfinishedFrame.incSampleCount();

                            sample = (short) normalize(unfinishedSample.getSample());
                            // calculate average sample
                            avg = 0.0001 * sample + 0.9999 * avg;
                            // use average to remove offset
                            sample = (short) (sample - avg);
                            // apply additional filtering if necessary
                            if (filters != null) sample = filters.apply(sample);

                            channels[channelCounter][sampleCounters[channelCounter]] = sample;

                            sampleCounters[channelCounter]++;

                            unfinishedSample = null;
                            if (unfinishedFrame.isFull()) {
                                //LOGD(TAG, " --> FRAME END");
                                unfinishedFrame = null;
                            }
                        } else {
                            msb = b1 & CLEANER;
                            // we already started the frame so if msb is greater then 127 drop whole frame
                            if (msb > 127) {
                                LOGW(TAG, "MSB > 127 WITHIN THE FRAME! DROP WHOLE FRAME!");
                                unfinishedFrame = null;
                                unfinishedSample = null;
                                channelCounter = 0;
                            } else {
                                channelCounter++;

                                //LOGD(TAG, " --> MSB " + (channelCounter + 1) + ". CHANNEL");
                                unfinishedSample = new Sample(msb);
                            }
                        }
                    } else {
                        msb = b1 & CLEANER;
                        if (msb > 127) {
                            channelCounter = 0;

                            //LOGD(TAG, " --> FRAME START");
                            unfinishedFrame = new Frame(tmpChannelCount);
                            //LOGD(TAG, " --> MSB " + (channelCounter + 1) + ". CHANNEL");
                            unfinishedSample = new Sample(msb);
                        } else {
                            LOGW(TAG, "MSB < 128 AT FRAME START! DROP!");
                        }
                    }
                }

                escapeSequence.reset();
            } else {
                if (escapeSequence.isCompleted()) {
                    // let's process incoming message
                    processEscapeSequenceMessage(escapeSequence.getMessage(), sampleCounters[CHANNEL_INDEX]);
                    // clear the escape sequence instance so we can start detecting the next one
                    LOGD(TAG, "Escape sequence is completed");
                    escapeSequence.reset();
                }
            }
        }

        //LOGD(TAG, "CHANNEL #1 SAMPLES COUNT: " + sampleCounters[0] + "/" + channels[0].length);
        //LOGD(TAG, "CHANNEL #2 SAMPLES COUNT: " + sampleCounters[1] + "/" + channels[1].length);

        currentChannel = channelCounter;
        average = avg;

        if (sampleCounters[CHANNEL_INDEX] == 0) {
            events.clear();
            return new short[0];
        }

        return Arrays.copyOfRange(channels[CHANNEL_INDEX], 0, sampleCounters[CHANNEL_INDEX]);
    }

    private int normalize(int sample) {
        return (sample - 512) * 30;
    }

    // Processes escape sequence message and triggers appropriate listener
    private void processEscapeSequenceMessage(byte[] messageBytes, int sampleIndex) {
        final String message = new String(messageBytes);
        LOGD(TAG, "ESCAPE MESSAGE: " + message);
        // check if it's board type message
        if (listener != null) {
            if (SampleStreamUtils.isHardwareTypeMsg(message)) {
                listener.onSpikerBoxHardwareTypeDetected(SampleStreamUtils.getBoardType(message));
            } else if (SampleStreamUtils.isSampleRateAndNumOfChannelsMsg(message)) {
                listener.onMaxSampleRateAndNumOfChannelsReply(SampleStreamUtils.getMaxSampleRate(message),
                    SampleStreamUtils.getChannelCount(message));
            } else if (SampleStreamUtils.isEventMsg(message)) {
                events.put(sampleIndex, SampleStreamUtils.getEventNumber(message));
            }
        }
    }

    /**
     * Represents single frame of a sample stream sent by BYB hardware.
     */
    private class Frame {

        int channelCount;
        int counter;

        Frame(int channelCount) {
            this.channelCount = channelCount;
        }

        boolean isFull() {
            return counter >= channelCount;
        }

        void incSampleCount() {
            counter++;
        }
    }

    /**
     * Represents single sample of a sample stream sent by BYB hardware.
     */
    private class Sample {

        int lsb, msb; // less and most significant bytes

        Sample(int msb) {
            this.msb = msb;
        }

        void setLsb(int lsb) {
            this.lsb = lsb;
        }

        int getSample() {
            msb = msb & REMOVER;
            msb = msb << 7;
            lsb = lsb & REMOVER;
            return msb | lsb;
        }
    }

    /**
     * Represents an escape sequence that holds different messages by sent by BYB hardware.
     */
    private class EscapeSequence {

        private final String TAG = makeLogTag(EscapeSequence.class);

        private final byte[] MESSAGE_START_SEQUENCE =
            new byte[] { (byte) 255, (byte) 255, (byte) 1, (byte) 1, (byte) 128, (byte) 255 };
        private final byte[] MESSAGE_END_SEQUENCE = new byte[] {
            (byte) 255, (byte) 255, (byte) 1, (byte) 1, (byte) 129, (byte) 255
        };
        // message cannot be longer than 64 bytes so sequence is sequence start + 64 + sequence end
        private final int MAX_SEQUENCE_LENGTH = MESSAGE_START_SEQUENCE.length + 64 + MESSAGE_END_SEQUENCE.length;

        int index = 0;
        int tmpIndex = 0;
        boolean started;
        boolean ended;
        byte[] start = new byte[MESSAGE_START_SEQUENCE.length];
        byte[] end = new byte[MESSAGE_END_SEQUENCE.length];
        byte[] sequence = new byte[MAX_SEQUENCE_LENGTH];

        EscapeSequence() {
            reset();
        }

        /**
         * Tests whether specified {@code b} byte, when appended to already passed bytes, makes a valid escape sequence
         * and it returns {@code true} if it does, {@code false} otherwise.
         */
        boolean test(byte b) {
            // if message is longer then 64 bytes reset
            if (index >= MAX_SEQUENCE_LENGTH) {
                LOGD(TAG, "Escape message longer than 64 bytes. RESET!!!");
                return false;
            }

            // and next byte to sequence
            sequence[index] = b;

            // detect sequence start
            if (!started) {
                if (MESSAGE_START_SEQUENCE[tmpIndex] == b && index < MESSAGE_START_SEQUENCE.length) {
                    LOGD(TAG, "Detected " + tmpIndex + " byte of escape sequence START!");

                    start[tmpIndex++] = b;

                    if (Arrays.equals(MESSAGE_START_SEQUENCE, start)) {
                        LOGD(TAG, "Escape sequence started!");

                        started = true;
                        tmpIndex = 0; // reset index, we need it for sequence end
                    }

                    index++;

                    return true;
                } else {
                    index++;

                    return false;
                }
            } else if (!ended) {
                // populate sequence end test array with last 6 bytes
                System.arraycopy(sequence, index - MESSAGE_END_SEQUENCE.length + 1, end, 0,
                    MESSAGE_END_SEQUENCE.length);

                if (Arrays.equals(MESSAGE_END_SEQUENCE, end)) {
                    LOGD(TAG, "Escape sequence ended!");

                    ended = true;
                }

                index++;

                return true;
            }

            index++;

            return false;
        }

        /**
         * Resets the instance and prepares it for new detection of next escape sequence.
         */
        void reset() {
            index = 0;
            tmpIndex = 0;
            started = false;
            ended = false;
            start = new byte[MESSAGE_START_SEQUENCE.length];
            end = new byte[MESSAGE_END_SEQUENCE.length];
            sequence = new byte[MAX_SEQUENCE_LENGTH];
        }

        /**
         * Whether escape sequence is complete. This means that it contained start and end sequences and the message
         * in between.
         */
        boolean isCompleted() {
            return started && ended;
        }

        /**
         * Returns raw sequence of bytes passed to {@link EscapeSequence}. Raw sequence should be queried if the
         * sequence wasn't completed ({@link #isCompleted()} will return {@code false}). In this case caller should
         * retrieve the raw sequence and process it as sample data.
         */
        byte[] getSequence() {
            return Arrays.copyOfRange(sequence, 0, index);
        }

        /**
         * Returns message sent by the USB device (excluding start and end of escape sequence). Message should be
         * queried only if the sequence was complete which can be checked by calling {@link #isCompleted()}. If message
         * was not complete, caller should call {@link #getSequence()} and process the sequence as sample data.
         *
         * @see #getSequence()
         */
        byte[] getMessage() {
            // don't returns start and end of sequence
            final int from = MESSAGE_START_SEQUENCE.length;
            final int to = index - MESSAGE_END_SEQUENCE.length;

            if (index <= MESSAGE_START_SEQUENCE.length + MESSAGE_END_SEQUENCE.length) return new byte[0];

            return Arrays.copyOfRange(sequence, from, to);
        }
    }
}

File: app/src/main/cpp/byb/SampleStreamProcessor.cpp
//
// Created by Tihomir Leka <tihomir at backyardbrains.com>
//

#include "SampleStreamProcessor.h"
#include "SampleStreamUtils.h"

namespace backyardbrains {

    namespace processing {

        const char *SampleStreamProcessor::TAG = "SampleStreamProcessor";

        const unsigned char SampleStreamProcessor::ESCAPE_SEQUENCE_START[] = {0xFF, 0xFF, 0x01, 0x01, 0x80, 0xFF};
        const unsigned char SampleStreamProcessor::ESCAPE_SEQUENCE_END[] = {0xFF, 0xFF, 0x01, 0x01, 0x81, 0xFF};

        SampleStreamProcessor::SampleStreamProcessor(backyardbrains::utils::OnEventListenerListener *listener)
                : Processor(DEFAULT_SAMPLE_RATE, DEFAULT_CHANNEL_COUNT, DEFAULT_BITS_PER_SAMPLE) {
            SampleStreamProcessor::listener = listener;
        }

        SampleStreamProcessor::~SampleStreamProcessor() = default;


        void SampleStreamProcessor::process(const unsigned char *inData, const int length, short **outSamples,
                                       int *outSampleCounts, int *outEventIndices, std::string *outEventLabels,
                                       int &outEventCount, const int channelCount, int hardwareType) {
//            batchCounter++;

            if (prevChannelCount != channelCount) { // number of channels changed during processing of previous batch
                frameStarted = false;
                sampleStarted = false;
                currentChannel = 0;
            }

            // init samples (by channels)
            for (int i = 0; i < channelCount; i++) {
                sampleCounters[i] = 0;
            }
            // init events
            eventCounter = 0;

            short sample;
            int sampleIndex;
            byte lsb; // last significant byte
            byte b; // temp variable to hold currently processed bytes
            unsigned char uc; // temp variable to hold currently processed bytes as unsigned char

            for (int i = 0; i < length; i++) {
                uc = inData[i];

                // and next byte to custom message sent by SpikerBox
                escapeSequence[escapeSequenceIndex++] = uc;

                if (insideEscapeSequence) { // we are inside escape sequence
                    sampleIndex = sampleCounters[currentChannel] == 0 ? 0 : sampleCounters[currentChannel] - 1;
                    if (eventMessageIndex >= EVENT_MESSAGE_LENGTH) { // event message shouldn't be longer then 64 bytes
                        auto *copy = new unsigned char[eventMessageIndex + 1];
                        std::copy(eventMessage, eventMessage + eventMessageIndex, copy);
                        copy[eventMessageIndex] = 0;
                        // let's process incoming message
                        hardwareType=processEscapeSequenceMessage(copy, sampleIndex);

                        delete[] copy;
                        reset();
                    } else if (ESCAPE_SEQUENCE_END[tmpIndex] == uc) {
                        tmpIndex++;
                        if (tmpIndex == ESCAPE_SEQUENCE_START_END_LENGTH) {
                            auto *copy = new unsigned char[eventMessageIndex + 1];
                            std::copy(eventMessage, eventMessage + eventMessageIndex, copy);
                            copy[eventMessageIndex] = 0;
                            // let's process incoming message
                            hardwareType=      processEscapeSequenceMessage(copy, sampleIndex);

                            delete[] copy;
                            reset();
                        }
                    } else {
                        eventMessage[eventMessageIndex++] = uc;
                    }
                }
                else {
                    if (ESCAPE_SEQUENCE_START[tmpIndex] == uc) {
                        tmpIndex++;
                        if (tmpIndex == ESCAPE_SEQUENCE_START_END_LENGTH) {
                            tmpIndex = 0; // reset index, we need it for sequence end
                            insideEscapeSequence = true;
                        }
                        continue;
                    }

                    auto *sequence = new unsigned char[escapeSequenceIndex];
                    std::copy(escapeSequence, escapeSequence + escapeSequenceIndex, sequence);
                    for (int j = 0; j < escapeSequenceIndex; j++) {
                        b = sequence[j];
                        // check if we have unfinished frame
                        if (frameStarted) {
                            // check if we have unfinished sample
                            if (sampleStarted) {
                                lsb = b & CLEANER;

                                // if less significant byte is also grater then 127 drop whole frame
                                if (lsb > 127) {
                                    __android_log_print(ANDROID_LOG_DEBUG, TAG, "LSB > 127! DROP WHOLE FRAME!");
                                    frameStarted = false;
                                    sampleStarted = false;
                                    currentChannel = 0;
                                    continue;
                                }

                                // get sample value from most and least significant bytes
                                msb = msb & REMOVER;
                                msb = msb << 7u;
                                lsb = lsb & REMOVER;
                                if(backyardbrains::utils::SampleStreamUtils::HUMAN_HARDWARE==hardwareType){
                                    sample = (short) (((msb | lsb) - 8192) );
                                }else{
                                    sample = (short) (((msb | lsb) - 512) * 30);
                                }

                                // calculate average sample
                                average = 0.0001 * sample + 0.9999 * average;
                                // use average to remove offset
                                sample = (short) (sample - average);

                                channels[currentChannel][sampleCounters[currentChannel]++] = sample;

                                sampleStarted = false;
                                if (currentChannel >= channelCount - 1) frameStarted = false;
                            } else {
                                msb = b & CLEANER;
                                // we already started the frame so if msb is greater then 127 drop whole frame
                                if (msb > 127) {
                                    __android_log_print(ANDROID_LOG_DEBUG, TAG,
                                                        "MSB > 127 WITHIN THE FRAME! DROP WHOLE FRAME!");

                                    frameStarted = false;
                                    sampleStarted = false;
                                    currentChannel = 0;
                                } else {
                                    currentChannel++;

                                    sampleStarted = true;
                                }
                            }
                        } else {
                            msb = b & CLEANER;
                            if (msb > 127) {
                                currentChannel = 0;

                                frameStarted = true;
                                sampleStarted = true;
                            } else {
                                __android_log_print(ANDROID_LOG_DEBUG, TAG, "MSB < 128 AT FRAME START! DROP!");
//                                if (batchCounter > 100 && !batchPrinted) {
//                                    __android_log_print(ANDROID_LOG_DEBUG, TAG, "PREV BATCH: (%d)", inDataPrevLength);
//                                    char tmp[6];
//                                    char prev[inDataPrevLength * 5 - 2];
//                                    prev[0] = 0;
//                                    for (int r = 0; r < inDataPrevLength; r++) {
//                                        if (r == 0) {
//                                            sprintf(tmp, "%u", inDataPrev[r]);
//                                        } else {
//                                            sprintf(tmp, " ,%u", inDataPrev[r]);
//                                        }
//                                        strcat(prev, tmp);
//                                    }
//                                    __android_log_print(ANDROID_LOG_DEBUG, TAG, "%s", prev);
//                                    __android_log_print(ANDROID_LOG_DEBUG, TAG, "NEW BATCH: (%d)", length);
//                                    char current[inDataPrevLength * 5 - 2];
//                                    current[0] = 0;
//                                    for (int r = 0; r < length; r++) {
//                                        if (r == 0) {
//                                            sprintf(tmp, "%u", inData[r]);
//                                        } else {
//                                            sprintf(tmp, " ,%u", inData[r]);
//                                        }
//                                        strcat(current, tmp);
//                                    }
//                                    __android_log_print(ANDROID_LOG_DEBUG, TAG, "%s", current);
//                                    __android_log_print(ANDROID_LOG_DEBUG, TAG,
//                                                        "------------------------------------- (%d)", i);
//
//                                    batchPrinted = true;
//                                }

                                frameStarted = false;
                                sampleStarted = false;
                                currentChannel = 0;
                            }
                        }
                    }

                    delete[] sequence;

                    reset();
                }
            }

//            std::copy(inData, inData + length, inDataPrev);
//            inDataPrevLength = length;

            bool avoidFilteringOfChannels = stopFilteringAfterChannelIndex >= 0;
            for (int i = 0; i < channelCount; i++) {
                // apply additional filtering if necessary
                if (avoidFilteringOfChannels && i <= stopFilteringAfterChannelIndex)
                    applyFilters(i, channels[i], sampleCounters[i]);
                outSamples[i] = new short[sampleCounters[i]];
                std::copy(channels[i], channels[i] + sampleCounters[i], outSamples[i]);
                outSampleCounts[i] = sampleCounters[i];
            }
            std::copy(eventIndices, eventIndices + eventCounter, outEventIndices);
            std::copy(eventLabels, eventLabels + eventCounter, outEventLabels);
            outEventCount = eventCounter;

            prevChannelCount = channelCount;
        }

        int SampleStreamProcessor::processEscapeSequenceMessage(unsigned char *messageBytes, int sampleIndex) {
            // check if it's board type message
            std::string message = reinterpret_cast<char *>(messageBytes);
            __android_log_print(ANDROID_LOG_DEBUG, TAG, "ESCAPE SEQUENCE MESSAGE %s AT %d", message.c_str(),
                                sampleIndex);
            int hardwareType=-1;
            if (backyardbrains::utils::SampleStreamUtils::isHardwareTypeMsg(message)) {
                hardwareType=backyardbrains::utils::SampleStreamUtils::getHardwareType(message);
                listener->onSpikerBoxHardwareTypeDetected(
                        hardwareType);
            } else if (backyardbrains::utils::SampleStreamUtils::isSampleRateAndNumOfChannelsMsg(message)) {
                const int sampleRate = backyardbrains::utils::SampleStreamUtils::getMaxSampleRate(message);
                const int channelCount = backyardbrains::utils::SampleStreamUtils::getChannelCount(message);
                listener->onMaxSampleRateAndNumOfChannelsReply(sampleRate, channelCount);
                setSampleRateAndChannelCount(sampleRate, channelCount);
            } else if (backyardbrains::utils::SampleStreamUtils::isEventMsg(message)) {
                eventIndices[eventCounter] = sampleIndex;
                eventLabels[eventCounter++] = backyardbrains::utils::SampleStreamUtils::getEventNumber(message);
            } else if (backyardbrains::utils::SampleStreamUtils::isExpansionBoardTypeMsg(message)) {
                const int expansionBoardType = backyardbrains::utils::SampleStreamUtils::getExpansionBoardType(message);
                listener->onExpansionBoardTypeDetection(expansionBoardType);
                updateProcessingParameters(expansionBoardType);
            } else if (backyardbrains::utils::SampleStreamUtils::isHumanSpikerBoxType300(message)) {
                const int boardState = backyardbrains::utils::SampleStreamUtils::getHumanSpikerBoxType300(message);
                listener->onHumanSpikerBoardState(boardState);
            } else if (backyardbrains::utils::SampleStreamUtils::isHumanSpikerBoxType300Audio(message)) {
                const int audioState = backyardbrains::utils::SampleStreamUtils::getHumanSpikerBoxType300Audio(message);
                listener->onHumanSpikerBoardAudioState(audioState);
            }
            return hardwareType;
        }

        void SampleStreamProcessor::updateProcessingParameters(int expansionBoardType) {
            switch (expansionBoardType) {
                default:
                case backyardbrains::utils::SampleStreamUtils::NONE_BOARD_DETACHED:
                    setSampleRateAndChannelCount(DEFAULT_SAMPLE_RATE, DEFAULT_CHANNEL_COUNT);
                    stopFilteringAfterChannelIndex = -1;
                    break;
                case backyardbrains::utils::SampleStreamUtils::ADDITIONAL_INPUTS_EXPANSION_BOARD:
                    setSampleRateAndChannelCount(EXPANSION_BOARDS_SAMPLE_RATE, ADDITIONAL_INPUTS_CHANNEL_COUNT);
                    stopFilteringAfterChannelIndex = 1;
                    break;
                case backyardbrains::utils::SampleStreamUtils::HAMMER_EXPANSION_BOARD:
                case backyardbrains::utils::SampleStreamUtils::JOYSTICK_EXPANSION_BOARD:
                    setSampleRateAndChannelCount(EXPANSION_BOARDS_SAMPLE_RATE, HAMMER_JOYSTICK_CHANNEL_COUNT);
                    stopFilteringAfterChannelIndex = 1;
                    break;
            }
        }

        void SampleStreamProcessor::reset() {
            insideEscapeSequence = false;
            tmpIndex = 0;
            escapeSequenceIndex = 0;
            eventMessageIndex = 0;
        }
    }
}